File: 3.longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        max_length = 0
        begin = -1
        char_dict = {}
        for i, c in enumerate(s):
            if c in char_dict and char_dict[c] > begin:
                begin = char_dict[c]
            char_dict[c] = i
            max_length = max(max_length, i - begin)
        return max_length

File: 3.longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int max_length = 0;
        int begin = -1;
        vector<int> map(256, -1);
        for (int end = 0; end < s.length(); end++) {
            if (map[s[end]] > begin) {
                begin = map[s[end]];
            }
            map[s[end]] = end;
            max_length = max(max_length, end - begin);
        }
        return max_length;
    }
};
